torch/csrc/jit/codegen/cuda/transform_replay.cpp

```diff
@@ -687,134 +687,370 @@ std::deque<TensorView*> producersFor(TensorView* tv) {
-  replayed_pos[producer_tv] = replayed_producer.second;
-  n_replayed_root_dims[producer_tv] = n_transformed_root_dims;
-
-  return true;
+  return result;
 }
 
-bool TransformPropagator::replayCasP(
-    TensorView* consumer_tv,
-    TensorView* producer_tv) {
-  if (consumer_tv == starting_tv) {
-    return false;
-  }
+// Given the preserved reference root ID info of a consumer, compute
+// the corresponding info in producer. The given info may be represented by
+// consumer's root domain, or rfactor domain, depending on how we reached the
+// consumer during propagation. If the given info is already represented with
+// consumer's root domain, then we directly map it to the producer's rfactor
+// domain. If the given info is represented with consumer's rfactor domain, we
+// need to first map it to the root domain of the consumer, then we can map it
+// to the producer's rfactor domain. The computed info will be represented by
+// rfactor domain as rfactor domain contains the raw information.
+std::vector<RootIDInfo> computeNextRootIDInfoPasC(
+    TensorView* producer,
+    TensorView* consumer,
+    const std::vector<RootIDInfo>& consumer_root_id_info) {
+  std::vector<RootIDInfo> result;
+  auto pairwise_map =
+      PairwiseRootDomainMap(producer, consumer /*, TODO: true or false?*/);
+  auto map = pairwise_map.mapConsumerToProducer(
+      consumer->domain(), producer->domain());
+
+  for (auto& info : consumer_root_id_info) {
+    RootIDInfo producer_info;
+    producer_info.is_complete = info.is_complete;
+    producer_info.is_rfactor = true;
+
+    // mapped rfactor ids in consumer -> mapped root ids in consumer
+    std::unordered_set<IterDomain*> root_ids;
+    if (info.is_rfactor && consumer->hasRFactor()) {
+      root_ids = mapRFactorToRoot(consumer, info.mapped_ids);
+    } else {
+      root_ids = info.mapped_ids;
+    }
 
-  auto producer_pos_it = replayed_pos.find(producer_tv);
-  if (producer_pos_it == replayed_pos.end()) {
-    return false;
-  }
+    // mapped root ids in consumer -> mapped rfactor ids in producer
+    for (auto id : root_ids) {
+      auto it = map.find(id);
+      if (it != map.end()) {
+        producer_info.starting_root_id = info.starting_root_id;
+        producer_info.mapped_ids.insert(it->second);
+      } else {
+        producer_info.is_complete = false;
+      }
+    }
 
-  auto pairwiseMap = PairwiseRootDomainMap(producer_tv, consumer_tv);
-  auto replayed_consumer = TransformReplay::replayCasP(
-      consumer_tv, producer_tv, producer_pos_it->second, pairwiseMap);
+    // We will stop at the rfactor ids in producer, and will not further map
+    // them into root ids in producer. This means, we only keep the unprocessed
+    // raw information of a tensor. This behavior is important to make sure that
+    // info is as accurate as possible throughout the propagation.
+    //
+    // For example, if we do a C->P->C propagation, we want to do
+    //   C(root) -> P(rfactor) -> C(root)
```

Comment:
Just want to be sure, here, the first `C` and the second `C` refer to different tensors, right?

Reply:
Exactly. I will use `C'` for the second one so people are not confused.

Reply:
done